beater/rl_cache.go

```diff
@@ -0,0 +1,88 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
+)
+
+const burstMultiplier = 5
+
+//The rlCache is a simple lru cache holding N=size rate limiter entities. Every
+//rate limiter entity allows N=rateLimit hits per key (=IP) per second, and has a
+//burst queue of limit*5.
+//As the used lru cache is of a fixed size, cache entries can get evicted, in
+//which case the evicted limiter is reused as the new rate limiter for the
+//current key. This adds a certain random factor to the rate limiting in case the
+//cache is full. The purpose is to avoid bypassing the rate limiting by sending
+//requests from cache_size*2 unique keys, which would lead to evicted keys and
+//the creation of new rate limiter entities with full allowance.
+type rlCache struct {
+	cache *simplelru.LRU
+	limit int
+
+	mu             sync.Mutex //guards limiter in cache
+	evictedLimiter *rate.Limiter
+}
+
+func NewRlCache(size, rateLimit int) (*rlCache, error) {
+	if size <= 0 || rateLimit < 0 {
+		return nil, errors.New("Cache initialization went wrong.")
```

Comment:
suggestion for more informative error message:
`cache initialization: size and rateLimit must be greater than zero`